java-checks/src/main/java/org/sonar/java/checks/LazyArgEvaluationCheck.java

```diff
@@ -0,0 +1,287 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherCollection;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Rule(key = "S2629")
+public class LazyArgEvaluationCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String SLF4J_LOGGER = "org.slf4j.Logger";
+  public static final String JUL_LOGGER = "java.util.logging.Logger";
+  private static final String STRING = "java.lang.String";
+
+  interface LogLevels {
+    List<MethodMatcher> log();
+    MethodMatcher test();
+
+    static MethodMatcher loggerProto(String typeDefinition) {
+      return MethodMatcher.create()
+        .typeDefinition(typeDefinition)
+        .addParameter(STRING);
+    }
+
+    static MethodMatcher levelTestProto(String typeDefinition) {
+      return MethodMatcher.create()
+        .typeDefinition(typeDefinition)
+        .withoutParameter();
+    }
+
+    static Stream<LogLevels> logLevels() {
+      return Stream.concat(Arrays.stream(SLF4J_LEVELS.values()), Arrays.stream(JUL_LEVELS.values()));
+    }
+
+    static List<MethodMatcher> varArgVariants(String logger, String methodName) {
+      return ImmutableList.of(loggerProto(logger).name(methodName),
+        loggerProto(logger).name(methodName).addParameter(TypeCriteria.anyType()),
+        loggerProto(logger).name(methodName).addParameter(TypeCriteria.anyType()).addParameter(TypeCriteria.anyType()));
+    }
+
+    enum SLF4J_LEVELS implements LogLevels {
+      TRACE,
+      DEBUG,
+      INFO,
+      WARN,
+      ERROR;
+
+      final List<MethodMatcher> log = varArgVariants(SLF4J_LOGGER, toString().toLowerCase(Locale.ROOT));
```

Comment:
The `varArgVariants` method is only called from here, so why not put it in this enum as private static method instead? And removing the toLowerCase from the method call to include in it the method body.